src/Agent.Plugins/PipelineArtifact/PipelineArtifactServer.cs

```diff
@@ -171,22 +171,40 @@ private BuildDropManager CreateBulidDropManager(AgentTaskPluginExecutionContext
             string[] minimatchFilters,
             CancellationToken cancellationToken)
         {
-            IDictionary<string, DedupIdentifier> artifactNameAndManifestId = new Dictionary<string, DedupIdentifier>();
-            foreach (var buildArtifact in buildArtifacts)
+            DownloadPipelineArtifactOptions options;
+            if (buildArtifacts.Count() > 1)
             {
+                IDictionary<string, DedupIdentifier> artifactNameAndManifestId = new Dictionary<string, DedupIdentifier>();
+                foreach (var buildArtifact in buildArtifacts)
+                {
+                    if (buildArtifact.Resource.Type != PipelineArtifactTypeName)
+                    {
+                        throw new ArgumentException("The artifact is not of the type Pipeline Artifact.");
+                    }
+                    artifactNameAndManifestId.Add(buildArtifact.Name, DedupIdentifier.Create(buildArtifact.Resource.Data));
+                }
+
+                // 2) download to the target path
+                options = DownloadPipelineArtifactOptions.CreateWithMultiManifestIds(
+                    artifactNameAndManifestId,
+                    targetDirectory,
+                    proxyUri: null,
+                    minimatchPatterns: minimatchFilters);
+            }
+            else
+            {
+                var buildArtifact = buildArtifacts.Single();
```

Comment:
SIngleOrDefault

Reply:
test for null

Reply:
We don't need to test for null, the client will throw if the buildArtifact is not found/doesn't exist